src/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs

```diff
@@ -597,6 +606,91 @@ private ProcessWaitState GetWaitState()
+
+            if (userId == null || groupId == null)
+            {
+                throw new Win32Exception(SR.Format(SR.UserDoesNotExist, startInfo.UserName));
+            }
+
+            return (userId.Value, groupId.Value);
+        }
+
+        private unsafe static (uint? userId, uint? groupId) GetUserAndGroupIds(string userName)
+        {
+            // First try with a buffer that should suffice for 99% of cases.
+            Interop.Sys.Passwd? passwd;
+            const int BufLen = 1024;
+            byte* stackBuf = stackalloc byte[BufLen];
+            if (TryGetPasswd(userName, stackBuf, BufLen, out passwd))
+            {
+                if (passwd == null)
+                {
+                    return (null, null);
+                }
+                return (passwd.Value.UserId, passwd.Value.GroupId);
+            }
+
+            // Fallback to heap allocations if necessary, growing the buffer until
+            // we succeed.  TryGetPasswd will throw if there's an unexpected error.
+            int lastBufLen = BufLen;
+            while (true)
+            {
+                lastBufLen *= 2;
+                byte[] heapBuf = new byte[lastBufLen];
+                fixed (byte* buf = &heapBuf[0])
+                {
+                    if (TryGetPasswd(userName, buf, heapBuf.Length, out passwd))
+                    {
+                        if (passwd == null)
+                        {
+                            return (null, null);
+                        }
+                        return (passwd.Value.UserId, passwd.Value.GroupId);
+                    }
+                }
+            }
+        }
+
+        private static unsafe bool TryGetPasswd(string name, byte* buf, int bufLen, out Interop.Sys.Passwd? passwd)
+        {
+            // Call getpwnam_r to get the passwd struct
+            Interop.Sys.Passwd tempPasswd;
+            int error = Interop.Sys.GetPwNamR(name, out tempPasswd, buf, bufLen);
+
+            // If the call succeeds, give back the passwd retrieved
+            if (error == 0)
+            {
+                passwd = tempPasswd;
+                return true;
+            }
+
+            // If the current user's entry could not be found, give back null,
+            // but still return true as false indicates the buffer was too small.
+            if (error == -1)
+            {
+                passwd = null;
+                return true;
+            }
+
+            var errorInfo = new Interop.ErrorInfo(error);
+
+            // If the call failed because the buffer was too small, return false to 
+            // indicate the caller should try again with a larger buffer.
+            if (errorInfo.Error == Interop.Error.ERANGE)
+            {
+                passwd = null;
+                return false;
+            }
+
+            // Otherwise, fail.
+            throw new IOException(errorInfo.GetErrorMessage(), errorInfo.RawErrno);
```

Comment:
>IOException [](start = 22, length = 11)

should we throw Win32Exception instead? I think we are throwing it on Linux too in some cases

Reply:
I copied the code from

https://github.com/dotnet/corefx/blob/17be4bc9b3fd9443026d2bf7d11a043b940bbcf4/src/System.Runtime.Extensions/src/System/Environment.Unix.cs#L429-L430
and
https://github.com/dotnet/corefx/blob/bfac45af5dddcbd70b6935fcf92b4b8755f3cd1c/src/Common/src/System/IO/PersistedFiles.Unix.cs#L159-L160

It seemed consistent to throw the same exception here.  But then looking at the Process.Start documentation, it lists Win32Exception and doesn't list IOException.  So I'll change to Win32Exception.